coverage.c

```diff
@@ -0,0 +1,590 @@
+/* coverage.c -- samtools coverage subcommand
+ 
+   Author: Florian P Breitwieser <florian.bw@gmail.com>
+
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.  */
+
+/* This program calculates coverage from multiple BAMs
+ * simutaneously, to achieve random access and to use the BED interface.
+ * To compile this program separately, you may:
+ *
+ *   gcc -g -O2 -Wall -o bamcov -D_MAIN_BAMCOV coverage.c -lhts -lz
+ */
+
+// C headers
+#include <config.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>  // variadic functions
+#include <limits.h>  // INT_MAX
+#include <math.h>    // round
+#include <stdbool.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+
+#include "htslib/sam.h"
+#include "htslib/hts.h"
+#include "samtools.h"
+#include "sam_opts.h"
+
+const char *VERSION = "0.1";
+
+typedef struct {  // auxiliary data structure to hold a BAM file
+    samFile *fp;     // file handle
+    bam_hdr_t *hdr;  // file header
+    hts_itr_t *iter; // iterator to a region - NULL for us by default
+    int min_mapQ;    // mapQ filter
+    int min_len;     // length filter
+    unsigned int n_reads;  // records the number of reads seen in file
+    unsigned int n_selected_reads; // records the number of reads passing filter
+    unsigned long summed_mapQ; // summed mapQ of all reads passing filter
+} bam_aux_t;
+
+typedef struct {  // auxiliary data structure to hold stats on coverage
+    unsigned long long n_covered_bases;
+    unsigned long long summed_coverage;
+    unsigned long long summed_baseQ;
+    unsigned long long summed_mapQ;
+    unsigned int n_reads;
+    unsigned int n_selected_reads;
+    int32_t tid;    // chromosome ID, defined by header
+    int beg;
+    int end;
+    int bin_width;
+} stats_aux_t;
+
+int read_file_list(const char *file_list, int *n, char **argv[]);
+
+static int usage(int status) {
+    fprintf(stderr, "\nUsage: samtools coverage [options] in1.bam [in2.bam [...]]\n"
```

Comment:
Ideally usage statements should go to stdout unless they are infact an error (eg someone tried "samtools coverage -x").  Possibly stdout always is good enough given the "coverage: invalid option -- 'x'" will be stderr anyway.

I know a lot of our other commands use stderr, but we also have this listed in the test harness as "known failures" for us to fix one day.

Reply:
Good point, made usage print to stdout and always return EXIT_SUCCESS.